### PR TITLE
Replaced HashMap in ZooCacheFactory with bounded-size Cache

### PR DESCRIPTION
**Pull request by dlmarion**
Related to #2699

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCacheFactory.java [hunk 1]
```diff
@@ -18,18 +18,17 @@
...
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+
 /**
  * A factory for {@link ZooCache} instances.
  */
 public class ZooCacheFactory {
-  // TODO: make this better - LRU, soft references, ...
-  private static Map<String,ZooCache> instances = new HashMap<>();
+  private static Cache<String,ZooCache> instances = Caffeine.newBuilder().maximumSize(5).build();
```
Comment (EdColeman) [inline core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCacheFactory.java:31]: Why a max size of 5?  What determines that number?

Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCacheFactory.java:31]: I don't know - I think the use-case for multiple instances in a VM is to support testing. Normally that number should be 1.  Does a caffeine cache pre-allocate entries up to max size on construction? If not, then a larger number could be appropriate?

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCacheFactory.java [hunk 2]
```diff
@@ -119,7 +118,8 @@ public ZooCache getNewZooCache(String zooKeepers, int sessionTimeout) {
    */
   void reset() {
     synchronized (instances) {
-      instances.clear();
+      instances.invalidateAll();
+      instances.cleanUp();
```
Comment (ben-manes) [inline core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCacheFactory.java:122]: fyi, `cleanUp` is implicit if clearing the cache so this is not needed.

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCacheFactory.java (hunk 1 above)
Comment (ben-manes) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCacheFactory.java:31]: It does not (`initialCapacity(c)` would do that).

**Review by ctubbsii — CHANGES_REQUESTED** [review]
I'm not opposed to using a better cache API, but I think there's a misunderstanding of what this is being used for.

It should not be an LRU cache. We are using this "cache" to track all the instances that have been created, so we can explicitly close them when the last legacy client has gone away. This is part of the "SingletonManager" code, and I anticipate removing ZooCacheFactory as soon as we can remove ZooKeeperInstance, hopefully in 3.0.

So, it might not be worth doing anything here other than removing the obsolete TODO. But, it definitely shouldn't be an LRU Cache... we need to track all created instances... not just the most recent ones.

**Comment by dlmarion** [comment]
Closing due to @ctubbsii comment

**Pull request closed without merging**